### PR TITLE
FINERACT-2067: NPE business date partition step COB

### PR DESCRIPTION
**Pull request by kulminsky**
## Description

This is to fix an NPE issue we are seeing in some of our test environments during COB in multi-tenant mode. 

Possible cause:

- When running in multi-tenant mode, ES and IT COBs execute at the same time.
- Both of these could somehow both be trying to access the business date value, causing NPE intermittently.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [x] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [x] Create/update unit or integration tests for verifying the changes made.

- [x] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [x] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [x] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

**Comment by galovics** [comment]
@kulminsky can you pls explain what was the exact issue you solved? I'm not sure I understand the purpose of the code changes yet. Thanks.

**Comment by adamsaghy** [comment]
@kulminsky Please make sure to follow the following format for PR title and commit message:
`FINERACT-XXXX: <short message>`

The ":" was missing from your actual title and commit message, but the above one is used to automatically pick up during a release and connect with the relevant jira ticket.

**Comment by kulminsky** [comment]
> @kulminsky Please make sure to follow the following format for PR title and commit message: `FINERACT-XXXX: <short message>`
> 
> The ":" was missing from your actual title and commit message, but the above one is used to automatically pick up during a release and connect with the relevant jira ticket.

Sorry, I saw the comment only after the commit. I will follow the rules in future commits.

**Comment by kulminsky** [comment]
> @kulminsky can you pls explain what was the exact issue you solved? I'm not sure I understand the purpose of the code changes yet. Thanks.

@galovics 
In general, this is to fix an NPE issue we are seeing in some of our test environments in multi-tenant mode. 
Potential fix steps:
1) Make partition step StepScoped.

2) This creates a new instance of the partition step, instead of the default singleton in Spring batch.

3) Copy values from job execution context into step execution context before partition step.

4) We then give the business date value to the step context before the step runs, ensuring other jobs/steps are not trying to access the business date value.

File: fineract-provider/src/main/java/org/apache/fineract/cob/loan/LoanCOBPartitioner.java [hunk 1]
```diff
@@ -58,10 +58,10 @@ public class LoanCOBPartitioner implements Partitioner {
 
     private final Long numberOfDays;
 
-    @Value("#{jobExecutionContext['BusinessDate']}")
+    @Value("#{stepExecutionContext['BusinessDate']}")
     @Setter
     private LocalDate businessDate;
-    @Value("#{jobExecutionContext['IS_CATCH_UP']}")
+    @Value("#{StepExecutionContext['IS_CATCH_UP']}")
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/cob/loan/LoanCOBPartitioner.java]: I think the S in the beginning should not be capitalized.

File: fineract-provider/src/main/java/org/apache/fineract/cob/listener/JobExecutionContextCopyListener.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import org.springframework.batch.core.StepExecutionListener;
+import org.springframework.batch.item.ExecutionContext;
+
+/**
+ * {@link StepExecutionListener} to copy values from Job execution context into Step execution context.
+ */
+@Slf4j
+public class JobExecutionContextCopyListener implements StepExecutionListener {
+
+    private static final List<String> STEP_EXECUTION_KEYS = List.of("BusinessDate", "IS_CATCH_UP");
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/cob/listener/JobExecutionContextCopyListener.java]: I'd actually not hardcode this. The listener seems to be very generic yet it has hardcoded keys related to the Loan COB.

Let's accept these keys as constructor parameters instead.

**Review by galovics — CHANGES_REQUESTED** [review]

File: fineract-provider/src/main/java/org/apache/fineract/cob/loan/LoanCOBPartitioner.java (hunk 1 above)
Comment (kulminsky) [reply fineract-provider/src/main/java/org/apache/fineract/cob/loan/LoanCOBPartitioner.java]: Done

File: fineract-provider/src/main/java/org/apache/fineract/cob/listener/JobExecutionContextCopyListener.java (hunk 2 above)
Comment (kulminsky) [reply fineract-provider/src/main/java/org/apache/fineract/cob/listener/JobExecutionContextCopyListener.java]: Done

File: fineract-provider/src/main/java/org/apache/fineract/cob/listener/JobExecutionContextCopyListener.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+    /**
+     * Method to copy values from Job execution context into Step execution context before step execution.
+     *
+     * @param stepExecution
+     *            the step to be executed.
+     */
+    @Override
+    public void beforeStep(final StepExecution stepExecution) {
+        log.info("Before step: copying job execution context to step [{}]", stepExecution.getStepName());
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/cob/listener/JobExecutionContextCopyListener.java]: Please change this to `debug` scope. It is usually not a useful information to log...

**Review by adamsaghy — APPROVED** [review]
LGTM

**Review by galovics — APPROVED** [review]

**Pull request merged**